File: 2024/01/main.ts
import { run } from '../shared/runtime.ts';

export type ParseOutput = ReturnType<typeof parse>;

export function parse(input: string) {
  return input
    .trim()
    .split('\n')
    .map(line => line.split(/\s+/).map(Number))
    .reduce(
      (acc, [c1, c2]) => [
        [...acc[0], c1],
        [...acc[1], c2],
      ],
      [[], []] as [number[], number[]]
    ) as [number[], number[]];
}

export function solvePart1(columns: ParseOutput) {
  const [left, right] = columns.map(column => column.sort((a, b) => a - b));
  return left.reduce((acc, val, i) => acc + Math.abs(right[i] - val), 0);
}

export function solvePart2([left, right]: ParseOutput) {
  const rightGroups = Object.groupBy(right, n => n);
  return left.reduce((acc, num) => acc + num * (rightGroups[num]?.length ?? 0), 0);
}

if (import.meta.main) {
  run(parse, solvePart1, solvePart2);
}

File: 2024/01/main.go
package main

import (
	"math"
	"slices"
	"strconv"
	"strings"

	"github.com/janvoracek/advent-of-code/aoc/2024/shared"
)

type ParseOutput = [2][]int

func parse(input string) ParseOutput {
	lines := strings.Split(strings.TrimSpace(input), "\n")
	result := ParseOutput{[]int{}, []int{}}
	for _, line := range lines {
		for i, part := range strings.Fields(line) {
			nr, _ := strconv.Atoi(part)
			result[i] = append(result[i], nr)
		}
	}
	return result
}

func solvePart1(input ParseOutput) any {
	slices.Sort(input[0])
	slices.Sort(input[1])
	sum := 0
	for i, v := range input[0] {
		sum += int(math.Abs(float64(v - input[1][i])))
	}
	return sum
}

func solvePart2(input ParseOutput) any {
	rightGroups := make(map[int]int)
	for _, num := range input[1] {
		rightGroups[num]++
	}

	sum := 0
	for _, num := range input[0] {
		sum += num * rightGroups[num]
	}

	return sum
}

func main() {
	shared.Run(parse, solvePart1, solvePart2)
}
